components/supervisor/pkg/serverapi/publicapi.go

```diff
@@ -0,0 +1,365 @@
+			"function:openPort",
+			"function:trackEvent",   // do we need this?
+			"function:getWorkspace", // do we need this?
+		},
+	})
+	if err != nil {
+		log.WithError(err).Error("cannot get token for Gitpod API")
+		return nil
+	}
+	// server api
+	gitpodService, err := gitpod.ConnectToServer(cfg.Endpoint, gitpod.ConnectToServerOpts{
+		Token: tknres.Token,
+		Log:   log.Log,
+		ExtraHeaders: map[string]string{
+			"User-Agent":              "gitpod/supervisor",
+			"X-Workspace-Instance-Id": cfg.InstanceID,
+			"X-Client-Version":        cfg.SupervisorVersion,
+		},
+	})
+	if err != nil {
+		log.WithError(err).Error("cannot connect to Gitpod API")
+		return nil
+	}
+
+	service := &Service{
+		token:         tknres.Token,
+		gitpodService: gitpodService,
+		cfg:           cfg,
+		experiments: experiments.NewCustomClient(configcat.Config{
+			SDKKey:       "gitpod",
+			BaseURL:      fmt.Sprintf("%s%s", os.Getenv("GITPOD_HOST"), "/configcat"),
+			PollInterval: time.Minute * 1,
+			HTTPTimeout:  3 * time.Second,
+		}),
+		publicApiMetrics: grpc_prometheus.NewClientMetrics(),
+	}
+
+	service.publicApiMetrics.EnableClientHandlingTimeHistogram(
+		// it should be aligned with https://github.com/gitpod-io/gitpod/blob/196a109eee50bfb7da2c6b858a3e78f2a2d0b26f/install/installer/pkg/components/ide-metrics/configmap.go#L199
+		grpc_prometheus.WithHistogramBuckets([]float64{.005, .025, .05, .1, .5, 1, 2.5, 5, 30, 60, 120, 240, 600}),
+	)
+
+	// public api
+	service.tryConnToPublicAPI()
+	// listen to server instance update
+	go service.listenInstanceUpdate(ctx, cfg.InstanceID)
+
+	// TODO: use public api?
+	if wsInfo, err := gitpodService.GetWorkspace(ctx, cfg.WorkspaceID); err != nil {
+		log.WithError(err).Error("cannot get workspace info")
+	} else {
+		service.ownerID = wsInfo.Workspace.OwnerID
+	}
+	return service
+}
+
+func (service *Service) tryConnToPublicAPI() {
+	endpoint := fmt.Sprintf("api.%s:443", service.cfg.Host)
+
+	log.WithField("endpoint", endpoint).Info("connecting to PublicAPI...")
+	opts := []grpc.DialOption{
+		grpc.WithTransportCredentials(credentials.NewTLS(&tls.Config{})),
+		grpc.WithUnaryInterceptor(grpc_middleware.ChainUnaryClient([]grpc.UnaryClientInterceptor{
+			service.publicApiMetrics.UnaryClientInterceptor(),
+			func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+				withAuth := metadata.AppendToOutgoingContext(ctx, "authorization", "Bearer "+service.token)
+				return invoker(withAuth, method, req, reply, cc, opts...)
+			},
+		}...)),
+	}
+	if conn, err := grpc.Dial(endpoint, opts...); err != nil {
+		log.WithError(err).Errorf("failed to dial public api %s", endpoint)
+	} else {
+		service.publicAPIConn = conn
+	}
+}
+
+func (s *Service) usePublicAPI(ctx context.Context) bool {
+	tmp := s.experiments.GetBoolValue(ctx, "supervisor_experimental_publicapi", false, experiments.Attributes{
+		UserID: s.ownerID,
+	})
```

Comment:
I'd recommend moving this into `experiments/flags.go` such that the check is in one place, and easily findable when you're removing a flag